### PR TITLE
add code.py

### PR DESCRIPTION
**Pull request by SzySteve**
# Code Review

#### Imagine that a junior developer has issued this pull request and requested that you review it and give feedback.

Please review for correctness, performance, and style.  The language is Python, but you do not need to know Python and you will not be judged based on your knowledge of the language.  Your interviewers can answer questions about Python.  Feel free to use any resources you need -- documentation, google, stackoverflow, etc.

## Background

Kinesis is a high-throughput stream buffer, similar to Kafka.  This PR contains a function to be run against one of our kinesis streams in an AWS Lambda function.  If data is available on the stream, the function will automatically be called and have access to an iterator over an arbitrary number of records (by default, up to 10,000) that were written to the stream.

**AWS Lambda functions in Python:** https://docs.aws.amazon.com/lambda/latest/dg/python-programming-model.html
**Tutorial on using Lambda with Kinesis:** https://docs.aws.amazon.com/lambda/latest/dg/with-kinesis-example-deployment-pkg.html

The AWS documentation provides this example to get you started:
```python
   def lambda_handler(event, context):
       for record in event['Records']:
          # Kinesis data is base64 encoded so decode here
          payload = base64.b64decode(record["kinesis"]["data"])
          print "Decoded payload: " + payload
```

Our function here will act on Monetate's session stream.  Each record contains data about the behaviour of a single visitor on one of our clients' sites.  Records in the stream are JSON format.  An example of a record (whitespace added):
```javascript
      {
        "accountId": 1256,
        "browser": "firefox",
        "debug": {
          "diagnostics": true,
          "serverId": "a-1d9a732ceb2a",
          "serverTime": 1501694454394
        },
        "guid": "387|2:1500924051963:705589531",
        "offers": {
          "431991-1": 1501694106,
          "493733-1": 1501694106,
          "813051-1": 1501694106
        },
        "os": "Windows 10",
        "productsViewed": [
          {
            "productId": "5350304",
            "time": "1501694294"
          },
          {
            "productId": "342934423",
            "time": "1501694333"
          },
          {
            "productId": "14029412",
            "time": "1501694312"
          },
          {
            "productId": "5350304-a",
            "time": "1501694214"
          }
        ],
        "timestamp": 1501694454,
        "urls": [
          "https://monetate-example.com/?ref=logo",
          "https://monetate-example.com/products/",
          "https://monetate-example.com/products/category/widgets/",
          "https://monetate-example.com/products/product/342934423",
          "https://monetate-example.com/products/product/14029412",
          "https://monetate-example.com/products/product/5350304",
          "https://monetate-example.com/cart/",
          "https://checkout.monetate-example.com/checkout/",
          "https://checkout.monetate-example.com/confirm/"
        ]
      }
```

The analytics database is AWS Aurora, which implements the same interface as MySQL.

Assume there exists a table called 'account_session_count' which has columns:
 - account_id INTEGER PRIMARY_KEY,
 - session_count INTEGER,
 - last_session_timestamp INTEGER

You can execute queries like this:
`db_engine.execute("SELECT * FROM account_session_count")`
SELECT statements will return a list of lists.  INSERT, UPDATE, and DELETE statements will return an integer number of rows affected.

### CODE REVIEW

File: code.py [hunk 1]
```diff
@@ -0,0 +1,56 @@
+
```
Comment (aelavender) [inline code.py]: test

File: code.py [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+    for record in event['Records']:
+        # Kinesis data is base64 encoded so decode here
+        payload = base64.b64decode(record["kinesis"]["data"])
+        data = json.loads(payload)
+
+        # Make timestamp human readable.  isoformat() returns a string like 'YYYY-MM-DDTHH:MM:SS.uuuuuuu'
+        data['timestamp'] = datetime.datetime.utcfromtimestamp(data['timestamp']).isoformat()
+
+        # If sessionizer was run in debug mode, the debug key appears.  Don't show it to end-users.
+        del data['debug']
```
Comment (chippermist) [inline code.py]: Should have a check before del as it will throw error

File: code.py [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+        # Append record to s3
+        s3_key = str(data['account_id']) + '/session-stream/' + data['timestamp'][:10].replace('-', '/')
+        s3_obj = S3_CLIENT.get_object(S3_BUCKET_NAME, s3_key)
+        file_contents = s3_obj['Body'].read()  # Download the s3 object and return contents as a string
+        file_contents += json.dumps(data) + '\n'
+        s3_obj.put(Body=file_contents)  # Upload the string back up to s3
+
+        # Increment session count
+        query_results = db_engine.execute(
+            "SELECT account_session_count FROM account_session_count WHERE account_id = " + str(data['accountId'])
```
Comment (chippermist) [inline code.py]: account_session_count is the 'table'. Column name for session count is 'session_count'

File: code.py [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+#
+# You can execute queries like this:
+#   db_engine.execute("SELECT * FROM account_session_count")
+
+
+S3_BUCKET_NAME = 'monetate-session-stream'
+S3_CLIENT = boto3.client('s3')
+
+
+def my_function(event, context):
```
Comment (chippermist) [inline code.py]: Purpose of the function

Explain the inputs to the function
--event 
--context

Explain output (if any)

File: code.py [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+# Assume there exists a table called 'account_session_count' which has columns:
+#   account_id INTEGER PRIMARY_KEY,
+#   session_count INTEGER,
+#   last_session_timestamp INTEGER
+#
+# You can execute queries like this:
+#   db_engine.execute("SELECT * FROM account_session_count")
+
+
+S3_BUCKET_NAME = 'monetate-session-stream'
```
Comment (chippermist) [inline code.py]: Explain the stream
--type of data
--contents

File: code.py [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+S3_CLIENT = boto3.client('s3')
+
+
+def my_function(event, context):
+    for record in event['Records']:
+        # Kinesis data is base64 encoded so decode here
+        payload = base64.b64decode(record["kinesis"]["data"])
+        data = json.loads(payload)
```
Comment (chippermist) [inline code.py]: Add comment -- parsing the decoded data into data stream

File: code.py [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+        # Include 'firstProductViewed'
+        data['firstProductViewed'] = data['productsViewed'][0]['productId']
+
+        # Product Management wanted to rename the key 'urls' to 'pagesViewed'
+        data['pagesViewed'] = data['urls']
+
+        # Append record to s3
+        s3_key = str(data['account_id']) + '/session-stream/' + data['timestamp'][:10].replace('-', '/')
+        s3_obj = S3_CLIENT.get_object(S3_BUCKET_NAME, s3_key)
+        file_contents = s3_obj['Body'].read()  # Download the s3 object and return contents as a string
```
Comment (chippermist) [inline code.py]: Check if the read was successful
--if it is then continue with next operation
--if not then return null (-1)

File: code.py [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+        file_contents += json.dumps(data) + '\n'
+        s3_obj.put(Body=file_contents)  # Upload the string back up to s3
```
Comment (chippermist) [inline code.py]: if != null then check

File: code.py [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+
+        # Increment session count
+        query_results = db_engine.execute(
```
Comment (chippermist) [inline code.py]: Add try-catch
Append the error to the catch, to inform user about the particular error.

File: code.py [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+            "SELECT account_session_count FROM account_session_count WHERE account_id = " + str(data['accountId'])
+        )
+        session_count = query_results[0][0]
+        db_engine.execute("UPDATE account_session_count SET session_count = " + str(session_count + 1) +
```
Comment (chippermist) [inline code.py]: Add try-catch
Check if successful and if not return the error code.

File: code.py [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+        s3_obj = S3_CLIENT.get_object(S3_BUCKET_NAME, s3_key)
+        file_contents = s3_obj['Body'].read()  # Download the s3 object and return contents as a string
+        file_contents += json.dumps(data) + '\n'
+        s3_obj.put(Body=file_contents)  # Upload the string back up to s3
+
+        # Increment session count
+        query_results = db_engine.execute(
+            "SELECT account_session_count FROM account_session_count WHERE account_id = " + str(data['accountId'])
+        )
+        session_count = query_results[0][0]
```
Comment (chippermist) [inline code.py]: Check the data is valid here before going to next step.

File: code.py [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+        db_engine.execute("UPDATE account_session_count SET session_count = " + str(session_count + 1) +
+                          ", last_timestamp = " + data['timestamp'] + " WHERE account_id = " + data['accountId'])
```
Comment (chippermist) [inline code.py]: last_timestamp does not exist
--maybe 'last_session_timestamp'

**Pull request closed without merging**